pkg/api/kptfile/v1alpha2/types.go

```diff
@@ -0,0 +1,213 @@
+type KptFile struct {
+	yaml.ResourceMeta `yaml:",inline"`
+
+	// UpstreamLock is a resolved locator for the last fetch of the package.
+	UpstreamLock *UpstreamLock `yaml:"upstreamLock,omitempty"`
+
+	// Info contains metadata such as license, documentation, etc.
+	Info *PackageInfo `yaml:"info,omitempty"`
+
+	// Subpackages declares the list of subpackages.
+	Subpackages []Subpackage `yaml:"subpackages,omitempty"`
+
+	// Pipeline declares the pipeline of functions.
+	Pipeline *Pipeline `yaml:"pipeline,omitempty"`
+
+	// Inventory contains parameters for the inventory object used in apply.
+	Inventory *Inventory `yaml:"inventory,omitempty"`
+}
+
+// OriginType defines the type of origin for a package.
+type OriginType string
+
+const (
+	// GitOrigin specifies a package as having been cloned from a git repository.
+	GitOrigin OriginType = "git"
+)
+
+// UpdateStrategyType defines the strategy for updating a package from upstream.
+type UpdateStrategyType string
+
+const (
+	// ResourceMerge performs a structural schema-aware comparison and
+	// merges the changes into the local package.
+	ResourceMerge UpdateStrategyType = "resource-merge"
+	// FastForward fails without updating if the local package was modified
+	// since it was fetched.
+	FastForward UpdateStrategyType = "fast-forward"
+	// ForceDeleteReplace wipes all local changes to the package.
+	ForceDeleteReplace UpdateStrategyType = "force-delete-replace"
+)
+
+// Upstream is a user-specified upstream locator for a package.
+type Upstream struct {
+	// Type is the type of origin.
+	Type OriginType `yaml:"type,omitempty"`
+
+	// Git is the locator for a package stored on Git.
+	Git *Git `yaml:"git,omitempty"`
+
+	// UpdateStrategy declares how a package will be updated from upstream.
+	UpdateStrategy UpdateStrategyType `yaml:"updateStrategy,omitempty"`
+}
+
+// UpstreamLock is a resolved locator for the last fetch of the package.
+type UpstreamLock struct {
+	// Type is the type of origin.
+	Type OriginType `yaml:"type,omitempty"`
+
+	// Git is the locator for a package stored on Git.
+	Git *Git `yaml:"git,omitempty"`
+}
+
+// Git contains information on the origin of packages fetched from a git repository.
+type Git struct {
+	// Repo is the git repository the package was cloned from.  e.g. https://
+	Repo string `yaml:"repo,omitempty"`
+
+	// Directory is the sub directory of the git repository that the package was cloned from
+	Directory string `yaml:"directory,omitempty"`
+
+	// Ref can be a Git branch, tag, or commit ID.
+	Ref string `yaml:"ref,omitempty"`
+
+	// Commit is the git commit that the package was fetched at
+	Commit string `yaml:"commit,omitempty"`
+}
+
+// PackageInfo contains information such as license, documentation, etc.
+// These fields are not consumed by any functionality in kpt and are simply passed through.
+type PackageInfo struct {
+	// URL is the location of the package.  e.g. https://github.com/example/com
+	URL string `yaml:"url,omitempty"`
```

Comment:
I think these comments points to why we might want to leave this undefined (at least for now) and allow users to extend the schema with their own metadata. There are lots of different information users can add to a package and it is not clear to me why these fields are the right ones. And if we want to let users extend it anyway, why not let them just define everything. If any of this information will be used by kpt directly, then I agree it should be in the schema.

Reply:
Renamed to Site. As discussed offline, we need to define the schema for Kptfile. There are extension points via `metadata.annotations|labels` and `packageInfo.keywords` for humans and automation to add arbitrary metadata. There are also some common metadata that every package manager needs. We want to define that here instead of having each user of kpt independently create a convention for it.